x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelInput.java

```diff
@@ -73,12 +96,99 @@ public boolean equals(Object o) {
+        public static final ConstructingObjectParser<InputObject, Void> STRICT_PARSER = createParser(false);
+
+        private static ConstructingObjectParser<InputObject, Void> createParser(boolean ignoreUnknownFields) {
+            ConstructingObjectParser<InputObject, Void> parser = new ConstructingObjectParser<>("trained_model_config_input_object",
+                ignoreUnknownFields,
+                a -> new InputObject((String) a[0], (String) a[1]));
+            parser.declareString(ConstructingObjectParser.constructorArg(), FIELD_NAME);
+            parser.declareString(ConstructingObjectParser.optionalConstructorArg(), FIELD_TYPE);
+            return parser;
+        }
+
+        public static InputObject fromXContent(XContentParser parser, boolean lenient) throws IOException {
+            return lenient ? LENIENT_PARSER.parse(parser, null) : STRICT_PARSER.parse(parser, null);
+        }
+
+        private final String fieldName;
+        private final String fieldType;
+
+        public InputObject(String fieldName) {
+            this(fieldName, null);
+        }
+
+        public InputObject(String fieldName, String fieldType) {
+            this.fieldName = ExceptionsHelper.requireNonNull(fieldName, FIELD_NAME);
+            this.fieldType = fieldType;
+        }
+
+        public InputObject(StreamInput in) throws IOException {
+            this.fieldName = in.readString();
+            if (in.getVersion().onOrAfter(Version.V_7_7_0)) {
+                this.fieldType = in.readOptionalString();
+            } else {
+                this.fieldType = null;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(fieldName, fieldType);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            TrainedModelInput.InputObject that = (TrainedModelInput.InputObject) o;
+            return Objects.equals(fieldName, that.fieldName)
+                && Objects.equals(fieldType, that.fieldType);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(fieldName);
+            if (out.getVersion().onOrAfter(Version.V_7_7_0)) {
+                out.writeOptionalString(fieldType);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(FIELD_NAME.getPreferredName(), fieldName);
+            if (fieldType != null) {
+                builder.field(FIELD_TYPE.getPreferredName(), fieldType);
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        public String getFieldName() {
+            return fieldName;
+        }
+
+        public String getFieldType() {
```

Comment:
The reason for keeping this a String, is for future bwc. Since we store the enumeration in the XContent object on disk, it might be a good idea to simply parse it out as a string.